store/tikv/client.go

```diff
@@ -172,6 +303,155 @@ func (a *connArray) Close() {
+				return
+			}
+			*entries = append(*entries, entry)
+			*requests = append(*requests, entry.req)
+		default:
+			return
+		}
+	}
+}
+
+// fetchMorePendingRequests fetches more pending requests from the channel.
+func fetchMorePendingRequests(
+	ch chan *batchCommandsEntry,
+	maxBatchSize int,
+	batchWaitSize int,
+	maxWaitTime time.Duration,
+	entries *[]*batchCommandsEntry,
+	requests *[]*tikvpb.BatchCommandsRequest_Request,
+) {
+	var waitStart, waitEnd time.Time
+	waitStart = time.Now()
+	waitEnd = waitStart
+
+	after := time.After(maxWaitTime)
+Loop1:
+	// Try to collect `batchWaitSize` requests, or wait `maxWaitTime`.
+	for len(*entries) < batchWaitSize {
+		select {
+		case entry := <-ch:
+			if entry == nil {
+				return
+			}
+			*entries = append(*entries, entry)
+			*requests = append(*requests, entry.req)
+		case now := <-after:
+			waitEnd = now
+			break Loop1
+		}
+	}
+
+	if waitEnd == waitStart {
+	Loop2:
+		// Do an additional non-block try.
+		for len(*entries) < maxBatchSize {
+			select {
+			case entry := <-ch:
+				if entry == nil {
+					return
+				}
+				*entries = append(*entries, entry)
+				*requests = append(*requests, entry.req)
+			default:
+				waitEnd = time.Now()
+				break Loop2
+			}
+		}
+	}
+	metrics.TiKVBatchWaitDuration.Observe(float64(waitEnd.Sub(waitStart)))
+}
+
+func (a *connArray) batchSendLoop(cfg config.TiKVClient) {
+	defer func() {
+		if r := recover(); r != nil {
+			buf := tidbutil.GetStack()
+			log.Errorf("gcWorker %v %s", r, buf)
+			metrics.PanicCounter.WithLabelValues(metrics.LabelBatchSendLoop).Inc()
+		}
+	}()
+
+	entries := make([]*batchCommandsEntry, 0, cfg.MaxBatchSize)
+	requests := make([]*tikvpb.BatchCommandsRequest_Request, 0, cfg.MaxBatchSize)
+	requestIDs := make([]uint64, 0, cfg.MaxBatchSize)
+
+	for {
+		// Choose a connection by round-robbin.
+		next := atomic.AddUint32(&a.index, 1) % uint32(len(a.v))
+		batchCommandsClient := a.batchCommandsClients[next]
+
+		entries = entries[:0]
+		requests = requests[:0]
+		requestIDs = requestIDs[:0]
+
+		metrics.TiKVPendingBatchRequests.Set(float64(len(a.batchCommandsCh)))
+		fetchAllPendingRequests(a.batchCommandsCh, int(cfg.MaxBatchSize), &entries, &requests)
+
+		if len(entries) < int(cfg.MaxBatchSize) && cfg.MaxBatchWaitTime > 0 {
+			tikvTransportLayerLoad := atomic.LoadUint64(batchCommandsClient.tikvTransportLayerLoad)
+			// If the target TiKV is overload, wait a while to collect more requests.
+			if uint(tikvTransportLayerLoad) >= cfg.OverloadThreshold {
+				fetchMorePendingRequests(
+					a.batchCommandsCh, int(cfg.MaxBatchSize), int(cfg.BatchWaitSize),
+					cfg.MaxBatchWaitTime, &entries, &requests,
+				)
+			}
+		}
+
+		length := len(requests)
+		maxBatchID := atomic.AddUint64(&batchCommandsClient.idAlloc, uint64(length))
+		for i := 0; i < length; i++ {
+			requestID := uint64(i) + maxBatchID - uint64(length)
+			requestIDs = append(requestIDs, requestID)
+			batchCommandsClient.batched.Store(requestID, entries[i])
```

Comment:
The modify of `batched` doesn't protected by `clientLock `. Will the newly added request be lost If there is a concurrent call of `failPendingRequests`?

Reply:
Nice catch. We can improve it. However there could still be some normal requests removed by `failPendingRequests`. Fortunately the upper layers can handle them correctly.